File: server/services/stewardship.ts
/**
 * Culture stewardship — "adopt a culture" ownership model (US-012).
 *
 * A steward adopts one or more **cultural domains** (a culture id, a region, or
 * an entity type). Stewards of a contribution's domain carry extra weight in the
 * multi-confirmation flow (`community-verification.ts`): their confirmation
 * lowers the required-reviewer bar, so domain expertise scales data quality.
 *
 * The model rules are **pure** (adopt / release / lookups take `now` as a
 * param); a thin `StewardshipStore` persists them as a single JSON file, mirroring
 * the JSON-on-disk pattern used by `contribution-service.ts` / `collections.ts`.
 */

import fs from "fs";
import path from "path";
import type { Contribution } from "./contribution-service";

// ============================================================================
// Model
// ============================================================================

export interface StewardAdoption {
  /** The steward (an opaque name/id — no auth). */
  steward: string;
  /** Normalized cultural-domain key (see `normalizeDomain`). */
  domain: string;
  adoptedAt: string;
  note?: string;
}

/** Normalize a domain string to a stable key (lowercase kebab). */
export function normalizeDomain(domain: string): string {
  return domain
    .trim()
    .toLowerCase()
    .replace(/[\s_]+/g, "-")
    .replace(/[^a-z0-9-]/g, "")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
}

/** Normalize a steward name for identity comparison (trim + lowercase). */
export function stewardKey(steward: string): string {
  return steward.trim().toLowerCase();
}

function sameAdoption(a: StewardAdoption, steward: string, domain: string): boolean {
  return stewardKey(a.steward) === stewardKey(steward) && a.domain === normalizeDomain(domain);
}

export type AdoptResult =
  | { adoptions: StewardAdoption[]; adoption: StewardAdoption; alreadyOwned: false }
  | { adoptions: StewardAdoption[]; adoption: StewardAdoption; alreadyOwned: true };

/**
 * Adopt a domain for a steward. Idempotent: re-adopting an owned domain returns
 * the existing adoption with `alreadyOwned: true` and leaves the list unchanged.
 * Pure.
 */
export function adoptDomain(
  adoptions: StewardAdoption[],
  input: { steward: string; domain: string; now: string; note?: string },
): AdoptResult {
  const domain = normalizeDomain(input.domain);
  const existing = adoptions.find((a) => sameAdoption(a, input.steward, domain));
  if (existing) {
    return { adoptions, adoption: existing, alreadyOwned: true };
  }
  const adoption: StewardAdoption = {
    steward: input.steward.trim(),
    domain,
    adoptedAt: input.now,
    note: input.note,
  };
  return { adoptions: [...adoptions, adoption], adoption, alreadyOwned: false };
}

/** Release a steward's claim on a domain. Pure. Returns whether anything changed. */
export function releaseDomain(
  adoptions: StewardAdoption[],
  steward: string,
  domain: string,
): { adoptions: StewardAdoption[]; released: boolean } {
  const next = adoptions.filter((a) => !sameAdoption(a, steward, domain));
  return { adoptions: next, released: next.length !== adoptions.length };
}

/** Stewards who have adopted the given domain (original casing). */
export function stewardsForDomain(adoptions: StewardAdoption[], domain: string): string[] {
  const key = normalizeDomain(domain);
  return adoptions.filter((a) => a.domain === key).map((a) => a.steward);
}

/** Domains a steward has adopted. */
export function domainsForSteward(adoptions: StewardAdoption[], steward: string): string[] {
  return adoptions.filter((a) => stewardKey(a.steward) === stewardKey(steward)).map((a) => a.domain);
}

/** True when `steward` owns `domain`. Pure. */
export function isStewardOf(
  adoptions: StewardAdoption[],
  steward: string,
  domain: string,
): boolean {
  return adoptions.some((a) => sameAdoption(a, steward, domain));
}

/**
 * Resolve the cultural domain a contribution belongs to, for stewardship
 * matching: an explicit `entityData.culturalDomain`, else the entity name
 * (its own culture, for a civilization/culture add), else the entity type. Pure.
 */
export function resolveContributionDomain(contribution: Contribution): string {
  const data = (contribution.entityData ?? {}) as Record<string, unknown>;
  const explicit = data.culturalDomain ?? data.cultureId ?? data.associatedCultureId;
  if (typeof explicit === "string" && explicit.trim()) return normalizeDomain(explicit);
  if (
    (contribution.entityType === "civilization") &&
    typeof data.name === "string" &&
    data.name.trim()
  ) {
    return normalizeDomain(data.name);
  }
  return normalizeDomain(contribution.entityType);
}

// ============================================================================
// Store
// ============================================================================

/**
 * Persists steward adoptions as a single JSON file (`stewards.json`) under an
 * injectable directory (default `data/stewardship`). Every mutation delegates to
 * the pure model functions above; the store owns only the fs + clock boundary.
 */
export class StewardshipStore {
  private dir: string;
  private file: string;

  constructor(dataDir: string = "data/stewardship") {
    this.dir = path.resolve(dataDir);
    this.file = path.join(this.dir, "stewards.json");
    if (!fs.existsSync(this.dir)) fs.mkdirSync(this.dir, { recursive: true });
  }

  private read(): StewardAdoption[] {
    if (!fs.existsSync(this.file)) return [];
    try {
      const parsed = JSON.parse(fs.readFileSync(this.file, "utf-8"));
      return Array.isArray(parsed) ? (parsed as StewardAdoption[]) : [];
    } catch {
      return [];
    }
  }

  private write(adoptions: StewardAdoption[]): void {
    fs.writeFileSync(this.file, JSON.stringify(adoptions, null, 2), "utf-8");
  }

  list(): StewardAdoption[] {
    return this.read();
  }

  listForDomain(domain: string): StewardAdoption[] {
    const key = normalizeDomain(domain);
    return this.read().filter((a) => a.domain === key);
  }

  adopt(input: { steward: string; domain: string; note?: string; now?: string }): AdoptResult {
    const now = input.now ?? new Date().toISOString();
    const result = adoptDomain(this.read(), { ...input, now });
    if (!result.alreadyOwned) this.write(result.adoptions);
    return result;
  }

  release(steward: string, domain: string): boolean {
    const { adoptions, released } = releaseDomain(this.read(), steward, domain);
    if (released) this.write(adoptions);
    return released;
  }

  isSteward(steward: string, domain: string): boolean {
    return isStewardOf(this.read(), steward, domain);
  }
}

File: services/api/src/pinakes/collab/stewardship.py
"""Culture stewardship — the "adopt a culture" ownership model.

The port of `server/services/stewardship.ts`. A steward claims a **cultural
domain** (a culture id, a region, or an entity type) and thereby carries extra
weight in the multi-confirmation flow: a steward of a contribution's domain can
verify it single-handedly, where three unrelated reviewers would otherwise be
needed.

Two things distinguish this store from its neighbours in :mod:`pinakes.collab`:

* **One file, not one per record.** Every adoption lives in a single
  `stewards.json` array, because the whole set is read on every question anyone
  asks of it ("is this reviewer a steward of that domain?") and it is small. The
  JSON-per-record :class:`~pinakes.collab.entities.RecordStore` would buy
  nothing here.
* **A record is identified by a pair, not an id.** ``(steward, domain)`` is the
  key: adopting is idempotent on it, and releasing is a filter over it. Steward
  identity compares case-insensitively (`stewardKey`) while the *stored* name
  keeps its original casing, so "Expert" and "expert" are one steward and the
  attribution still reads the way they typed it.

Every rule is a pure function over a list of adoptions, taking ``now`` as an
argument; :class:`StewardshipStore` only supplies the filesystem and the clock.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, NamedTuple

from pinakes.contributions.store import iso_now
from pinakes.paths import stewardship_dir

#: One adoption, as it is stored and served. A plain dict for the same reason
#: the other collaborative records are: the on-disk shape is the TypeScript
#: writer's, and both servers read this file during the cutover.
Adoption = dict[str, Any]

#: The single file every adoption lives in, under the stewardship directory.
STEWARDS_FILE = "stewards.json"

_SEPARATORS = re.compile(r"[\s_]+")
_DISALLOWED = re.compile(r"[^a-z0-9-]")
_RUNS = re.compile(r"-+")


def normalize_domain(domain: str) -> str:
    """A domain string reduced to its stable key: lowercase kebab.

    "Roman Empire", "roman_empire" and "  ROMAN-EMPIRE  " are one domain. The
    normalized form is what is *stored*, so a claim made under one spelling is
    found under any of them.
    """
    text = _SEPARATORS.sub("-", domain.strip().lower())
    text = _DISALLOWED.sub("", text)
    return _RUNS.sub("-", text).strip("-")


def steward_key(steward: str) -> str:
    """A steward name reduced for identity comparison. Never stored."""
    return steward.strip().lower()


def _same(adoption: Adoption, steward: str, domain: str) -> bool:
    """Is *adoption* this steward's claim on this domain?"""
    return steward_key(str(adoption.get("steward", ""))) == steward_key(
        steward
    ) and adoption.get("domain") == normalize_domain(domain)


class AdoptResult(NamedTuple):
    """The outcome of an adoption: what to persist, and what to answer with."""

    adoptions: list[Adoption]
    adoption: Adoption
    already_owned: bool


def adopt_domain(
    adoptions: list[Adoption],
    *,
    steward: str,
    domain: str,
    now: str,
    note: str | None = None,
) -> AdoptResult:
    """Claim a domain for a steward. Pure.

    **Idempotent**: re-adopting an owned domain returns the *existing* adoption
    with ``already_owned`` set and the list untouched, so a double-click does not
    mint a second claim with a later `adoptedAt`.
    """
    normalized = normalize_domain(domain)
    for existing in adoptions:
        if _same(existing, steward, normalized):
            return AdoptResult(adoptions, existing, True)

    adoption: Adoption = {
        "steward": steward.strip(),
        "domain": normalized,
        "adoptedAt": now,
    }
    # Absent, not null: `JSON.stringify` writes no key for the `undefined` the
    # TypeScript builder leaves, and both servers read this file.
    if note is not None:
        adoption["note"] = note
    return AdoptResult([*adoptions, adoption], adoption, False)


def release_domain(
    adoptions: list[Adoption], steward: str, domain: str
) -> tuple[list[Adoption], bool]:
    """Drop a steward's claim on a domain. Returns the list and whether it moved."""
    remaining = [a for a in adoptions if not _same(a, steward, domain)]
    return remaining, len(remaining) != len(adoptions)


def stewards_for_domain(adoptions: list[Adoption], domain: str) -> list[str]:
    """Steward names holding *domain*, in their original casing."""
    key = normalize_domain(domain)
    return [str(a["steward"]) for a in adoptions if a.get("domain") == key]


def domains_for_steward(adoptions: list[Adoption], steward: str) -> list[str]:
    """Normalized domains a steward holds."""
    key = steward_key(steward)
    return [
        str(a["domain"])
        for a in adoptions
        if steward_key(str(a.get("steward", ""))) == key
    ]


def is_steward_of(adoptions: list[Adoption], steward: str, domain: str) -> bool:
    """Does this steward hold this domain? The question the confirm flow asks."""
    return any(_same(a, steward, domain) for a in adoptions)


class StewardshipStore:
    """The filesystem half: one `stewards.json` under an injectable directory."""

    def __init__(self, directory: Path) -> None:
        self.directory = Path(directory)
        self.file = self.directory / STEWARDS_FILE

    def read(self) -> list[Adoption]:
        """Every adoption. A missing, unparseable or non-array file is empty.

        Unlike the JSON-per-record stores this **swallows** a malformed file,
        because the TypeScript did: a broken `stewards.json` degrades to "nobody
        is a steward", which costs a steward their shortcut rather than costing
        every reviewer the endpoint. Losing a *collection* silently would be
        worse than a 500; losing a claim that can be re-made is not.
        """
        if not self.file.is_file():
            return []
        try:
            parsed = json.loads(self.file.read_text(encoding="utf-8"))
        except ValueError:
            return []
        return parsed if isinstance(parsed, list) else []

    def _write(self, adoptions: list[Adoption]) -> None:
        self.directory.mkdir(parents=True, exist_ok=True)
        self.file.write_text(json.dumps(adoptions, indent=2), encoding="utf-8")

    def list_all(self) -> list[Adoption]:
        """Every adoption, in the order they were made."""
        return self.read()

    def list_for_domain(self, domain: str) -> list[Adoption]:
        """Adoptions of one domain, matched on its normalized key."""
        key = normalize_domain(domain)
        return [a for a in self.read() if a.get("domain") == key]

    def adopt(
        self,
        *,
        steward: str,
        domain: str,
        note: str | None = None,
        now: str | None = None,
    ) -> AdoptResult:
        """Claim a domain, persisting only when something actually changed."""
        result = adopt_domain(
            self.read(),
            steward=steward,
            domain=domain,
            note=note,
            now=now if now is not None else iso_now(),
        )
        if not result.already_owned:
            self._write(result.adoptions)
        return result

    def release(self, steward: str, domain: str) -> bool:
        """Drop a claim. Returns whether there was one to drop."""
        remaining, released = release_domain(self.read(), steward, domain)
        if released:
            self._write(remaining)
        return released

    def is_steward(self, steward: str, domain: str) -> bool:
        return is_steward_of(self.read(), steward, domain)


def store(directory: Path | None = None) -> StewardshipStore:
    """The stewardship store this process serves.

    Built per call from :func:`pinakes.paths.stewardship_dir`, which re-reads its
    environment override every time — the same no-singleton rule the rest of the
    service follows, and for the same two reasons: Express still reads this file
    for the confirm flow, and a test redirects it by setting one variable.
    """
    return StewardshipStore(
        directory if directory is not None else stewardship_dir()
    )
